cpp/src/sort/sort.cuh

```diff
@@ -0,0 +1,70 @@
+// -*-c++-*-
+
+#pragma once
+
+#include <cudf/types.h>
+
+namespace cusort {
+
+  /**
+   * @brief  Sort key value pairs distributed across multiple GPUs
+   *
+   *   This sort function takes arrays of keys and values distributed
+   *   around multiple GPUs 
+   *
+   * @param[in]   input_keys_d     The unsorted keys, stored in device arrays.
+   *                               input_keys_d[i] is the array of keys on GPU i
+   * @param[in]   input_values_d   The unsorted values, stored in device arrays.
+   *                               input_values_d[i] is the array of values on GPU i
+   * @param[in]   input_length_h   Host array containing the number of elements on
+   *                               each GPU in the input key/value arrays.
+   * @param[out]  output_keys_d    The sorted keys, stored in device arrays.
+   *                               output_keys_d[i] is the array of keys on GPU i
+   * @param[out]  output_values_d  The sorted values, stored in device arrays.
+   *                               output_values_d[i] is the array of values on GPU i
+   * @param[out]  output_length_h  Host array containing the number of elements on
+   *                               each GPU in the output key/value arrays.
+   * @param[in]   num_gpus         The number of GPUs
+   *
+   * @return   GDF_SUCCESS upon successful completion
+   */
+  template <typename Key_t, typename Value_t, typename Length_t>
```

Comment:
We could hard-code size_t.  But that would unnecessarily restrict the caller.  If they had an array of int32_t that contained the sizes they would need to create an array of size_t and copy the data.  I templated it to avoid that extra work.

The renumber function is also templated.  When I integrated it into a python binding I simply passed the correct arguments into the function call.

Reply:
For the initial use case (Coo2Csr), we probably don't need templating as much as we have the exact data types. In the longer term, we may want to change the key and value types.

Reply:
Sounds good